Added padding_idx=None option and new test cases for aten_embedding_bag

crypto-a opened the pull request:
## Fix Issues #2219, #2385 and the first part of #2489

This commit adds new test cases and the necessary implementation changes to correctly support the `padding_idx=None` option in the `aten_embedding_bag` operator. This aligns the ONNX Script operator with PyTorch's native behavior and expands test coverage for this feature.

**Key Changes:**

* **`core.py`**: The `aten_embedding_bag_padding_idx` function has been updated to handle `padding_idx=None`. This new code routes the operation to the standard `aten_embedding_bag` implementation when no padding indices are specified.
* **`extra_opinfo.py`**: Two new `OpInfo` definitions, `test_embedding_bag_with_padding_idx_none` and `test_embedding_bag_with_padding_idx_int`, have been added to the `OP_DB` list. These provide input samples to test the new and existing `padding_idx` functionality.
* **`ops_test_data.py`**: The `TESTED_TORCHLIB_OPS` tuple has been updated to include the new tests, ensuring they are discovered and executed by the test runner.

justinchuby commented on tests/function_libs/torch_lib/ops_test_data.py:
```diff
@@ -185,6 +185,24 @@
+) -> tuple[list[Any], dict[str, Any]]:
+    # ONNX attributes cannot be None; omit padding_idx if it’s None.
+    padding_idx = kwargs.pop("padding_idx", "___MISSING___")
+    if padding_idx is not "___MISSING___":
```
You can just do a membership check?

crypto-a replied:
replaced it with `padding_idx != "___MISSING___":`

crypto-a commented on onnxscript/function_libs/torch_lib/ops/core.py:
```diff
@@ -3101,27 +3101,44 @@
-        padding_idx = weight.shape[0] + padding_idx
+    if padding_idx is not None:
+        # Call the existing function for handling padding_idx
+        result, offset2bag, bag_size, max_indices =_aten_embedding_bag_1d_padding_idx_onnx(
```
removed extra arguments

crypto-a commented on onnxscript/function_libs/torch_lib/ops/core.py:
```diff
@@ -3101,27 +3101,44 @@
+        return result, offset2bag, bag_size, max_indices
+
+    # When padding_idx is None, use the standard embedding_bag implementation
+    result, offset2bag, bag_size, max_indices = _aten_embedding_bag_onnx(
```
removed extra arguments

crypto-a commented on tests/function_libs/torch_lib/ops_test_data.py:
```diff
@@ -185,6 +185,24 @@
+def _embedding_bag_input_wrangler(
+    args: list[Any], kwargs: dict[str, Any]
+) -> tuple[list[Any], dict[str, Any]]:
+    # ONNX attributes cannot be None; omit padding_idx if it’s None.
```
fixed this lint issue: replaced `’` with `'`

crypto-a commented on tests/function_libs/torch_lib/extra_opinfo.py:
```diff
@@ -2210,6 +2210,44 @@ def __init__(self):
+        ],
+    ),
+    opinfo_core.OpInfo(
+        "test_embedding_bag_with_padding_idx_int",
```
replaced the names with `ops.aten.embedding_bag.padding_idx_none` and `ops.aten.embedding_bag.padding_idx_int`

crypto-a commented on tests/function_libs/torch_lib/extra_opinfo.py:
```diff
@@ -2210,6 +2210,44 @@ def __init__(self):
+    opinfo_core.OpInfo(
+        "test_embedding_bag_with_padding_idx_int",
+        op=torch.nn.functional.embedding_bag,
+        dtypes=(torch.float32,),
```
changed it to `common_dtype.floating_types_and_half()`, similar to other test cases

jsmonson commented:
@crypto-a I would love to see this PR go in! Are you still planning on contributing it?

justinchuby commented:
Changes LGTM. CI is reporting

`TypeError: aten_embedding_bag() got an unexpected keyword argument 'padding_idx'`

justinchuby commented:
@crypto-a could you take a look at the CI errors and rebase from main? Thanks

crypto-a commented:
@justinchuby , I’ve fixed the CI issues. I had to remove the test cases I added myself. Turns out the scenarios I needed for my code to work were already covered in the OpInfo tests. All tests are now passing on my end. Could you please rerun CI? Thanks!

crypto-a commented:
Tests are passing, the branch is up to date, and it’s ready to merge

justinchuby commented on onnxscript/function_libs/torch_lib/ops/core.py:
```diff
@@ -2861,6 +2861,7 @@ def aten_embedding_bag(
     sparse: bool = False,
     per_sample_weights: Optional[TFloat] = None,
     include_last_offset: bool = False,
+    padding_idx: Optional[int] = None,
```
I checked https://github.com/pytorch/pytorch/blob/8656dea039bd0a31952a6a9792566e70b07429dc/aten/src/ATen/native/native_functions.yaml#L2372-L2378 and found that embedding_bag does not have padding_idx as a parameter. I think you only need to update aten_embedding_bag_padding_idx
